### ISSUE TITLE
How to exclude a variable from training, but not to remove it from the training set

### ISSUE THREAD

**Issue by rczen** [tidymodels/recipes#392]
Hi,
I need your advice.

How to exclude a variable from training, but not to remove it from the training set,
tried update_role() but doesn't seem to do what I wanted.

Thanks!

**Comment by topepo** [comment]
Can you please provide a minimal reprex (reproducible example)? The goal of a reprex is to make it as easy as possible for me to recreate your problem so that I can fix it: please help me help you!

If you've never heard of a reprex before, start by reading "[What is a reprex](https://github.com/tidyverse/reprex#what-is-a-reprex)", and follow the advice further down that page.

**Comment by rczen** [comment]
Let me ask it differently.
I am looking for a way to tell a recipe not to treat a variable as a predictor, and at the same time keep it as part of the baked dataset.

The reason for it that usually, we would not like to treat id's as predictors, but we want to keep them as part of the baked dataset since they are needed upon submission of prediction results. most common case in all ML competition.

**Comment by natejessee** [comment]
@rczen your initial guess that `update_role()` is what you're after was correct. You need to specify `update_role(id_variable_name, new_role = "id variable")`

If you look in the 'Examples' section in `help(update_role)` you can see that this changes the role of the 'sample' column (which is being treated as the id) from 'predictor' to 'id variable':

``` r
library(recipes)
#> Loading required package: dplyr
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
#> 
#> Attaching package: 'recipes'
#> The following object is masked from 'package:stats':
#> 
#>     step
data(biomass)

# Using the formula method, roles are created for any outcomes and predictors:
recipe(HHV ~ ., data = biomass) %>%
  summary()
#> # A tibble: 8 x 4
#>   variable type    role      source  
#>   <chr>    <chr>   <chr>     <chr>   
#> 1 sample   nominal predictor original
#> 2 dataset  nominal predictor original
#> 3 carbon   numeric predictor original
#> 4 hydrogen numeric predictor original
#> 5 oxygen   numeric predictor original
#> 6 nitrogen numeric predictor original
#> 7 sulfur   numeric predictor original
#> 8 HHV      numeric outcome   original

# However `sample` and `dataset` aren't predictors. Since they already have
# roles, `update_role()` can be used to make changes:
recipe(HHV ~ ., data = biomass) %>%
  update_role(sample, new_role = "id variable") %>%
  update_role(dataset, new_role = "splitting variable") %>%
  summary()
#> # A tibble: 8 x 4
#>   variable type    role               source  
#>   <chr>    <chr>   <chr>              <chr>   
#> 1 sample   nominal id variable        original
#> 2 dataset  nominal splitting variable original
#> 3 carbon   numeric predictor          original
#> 4 hydrogen numeric predictor          original
#> 5 oxygen   numeric predictor          original
#> 6 nitrogen numeric predictor          original
#> 7 sulfur   numeric predictor          original
#> 8 HHV      numeric outcome            original
```

<sup>Created on 2019-09-24 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by rczen** [comment]
Thanks, Nate, 
Indeed the recipe identifies it as a non-predictor variable.
but use it as a predictor at training time when using baked training set.
It seems that using caret with a recipe for fitting a model, this systematics is being considered.
I think it is a requirement that can't be fulfilled when baking the training since there is no mechanism to let the train not to consider non-predictors

**Comment by natejessee** [comment]
Ronen,

Can you provide an example on a small dataset (aka a "reprex") like Max suggested?

If I continue the previous example by training a model with caret the id column does _not_ get considered as a predictor:

``` r
library(plyr)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:plyr':
#> 
#>     arrange, count, desc, failwith, id, mutate, rename, summarise,
#>     summarize
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(gbm)
#> Loaded gbm 2.1.5
library(recipes)
#> 
#> Attaching package: 'recipes'
#> The following object is masked from 'package:stats':
#> 
#>     step
library(caret)
#> Loading required package: lattice
#> Loading required package: ggplot2

train <- biomass

rec <- recipe(HHV ~ ., data = train) %>%
  update_role(sample, new_role = "id variable") %>%
  update_role(dataset, new_role = "splitting variable")

cross_validation <- trainControl(method = "cv",
                                 number = 5)

model <- train(rec,
               train,
               method = "gbm",
               trControl = cross_validation,
               tuneLength = 1,
               verbose = FALSE)

predictors(model)
#> [1] "carbon"   "hydrogen"
model$finalModel$var.names
#> [1] "carbon"   "hydrogen" "oxygen"   "nitrogen" "sulfur"
varImp(model)
#> gbm variable importance
#> 
#>           Overall
#> carbon   100.0000
#> hydrogen   0.6384
#> nitrogen   0.0000
#> sulfur     0.0000
#> oxygen     0.0000
```

<sup>Created on 2019-09-24 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by rczen** [comment]
Ok your comments were helpful, I solved the issue by excluding the two non-predictors. while training on the baked version of the training data.

library(gbm)
#> Loaded gbm 2.1.5
library(recipes)
#> Loading required package: dplyr
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
#> 
#> Attaching package: 'recipes'
#> The following object is masked from 'package:stats':
#> 
#>     step
library(caret)
#> Loading required package: lattice
#> Loading required package: ggplot2

train <- biomass

baked_train <- recipe(HHV ~ ., data = train) %>%
  update_role(sample, new_role = "id variable") %>%
  update_role(dataset, new_role = "splitting variable") %>% 
  prep(train) %>% 
  bake(train)

cross_validation <- trainControl(method = "cv",
                                 number = 5)

model <- train(HHV ~ . -sample -dataset,
               baked_train,
               method = "gbm",
               trControl = cross_validation,
               tuneLength = 1,
               verbose = FALSE)

varImp(model)
#> gbm variable importance
#> 
#>           Overall
#> carbon   100.0000
#> hydrogen   0.7831
#> sulfur     0.0000
#> oxygen     0.0000
#> nitrogen   0.0000
Created on 2019-09-24 by the reprex package (v0.3.0)

**Comment by natejessee** [comment]
Glad you were able to figure it out.

Just an FYI that when you use caret you don't need to prep/bake the data, or manually deselect your columns from the training data--recipes will handle all of that for you. Instead of using `train(target ~., data)` (formula interface) you can use the [recipes interface](https://topepo.github.io/caret/using-recipes-with-train.html) which you do by providing a recipe object (eg `train(recipe, data)`).

When caret sees that you've provided an object whose class is `"recipe"` it will automatically handle the baking for you, and only include predictors whose role is set to `'predictor'`.

**Comment by rczen** [comment]
You are very kind,
Thank you.
